module4/lessons/Week4/Interfaces.md

```diff
@@ -0,0 +1,195 @@
+
+✅ CHECKPOINT ✅: Before continuing, work together to write a Slack message to your instructor explaining interfaces in your own words. Imagine your audience was you from last week and isn't familiar with the topic. Create your own example to use in the explanation. I recommend using a repl to code your example and try things out.
+
+Once you get the go-ahead from your instructor, move on to thinking about and researching why this might be helpful.
+
+1. What are some of the benefits of using interfaces?
+2. What might be some of the disadvantages of using interfaces?
+
+Next, look for examples of where you have used interfaces previously in this course. Because we already talked about `IEnumarable`, look for a different example. There are multiple!
+
+*Hint: Interfaces usually start with the work `I` and in Visual Studio you can hover over something you think might be an interface to learn more.*
+
+1. How is an interface used here?
+1. Why is an interface used here?
+
+✅ CHECKPOINT ✅: Before continuing, work together to write a Slack message to your instructor explaining the example of an interface that you found. Be sure to include how and why an interface is used here.
+
+Finally, move on to digging into at least one more advanced topic. Here are some ideas, or feel free to dig into your own questions!
+
+1. What happens if a class doesn't implement all the methods of an interface in C#?
+1. What are some of the most commonly used interfaces in C# and .NET?
+1. Can a class implement multiple interfaces? If yes, how can you do that?
+1. How do you use an interface to achieve polymorphism in C#? (You will also need to google the term `polymorphism` for this one!)
+1. Can you define properties in an interface with different access levels (e.g., public get and private set)?
+1. Can you provide default implementations for interface methods in C#?
+1. Does an interface have a constructor? Why or why not?
+1. What is an abstract class? How is it different from an interface?
+
+## Part 3: Presentation Requirements
+
+Each pair will have 7 minutes to present, and each partner should present for roughly equal time. Classmates will then have a couple of minutes to ask questions.
+
+1. Share at least one benefit of using interfaces and one potential disadvantage.
+1. Share one example of an interface from a past MVC project. Explain how and why an interface is used here.
+1. Explain one advanced part of interfaces you learned about today that was not covered in this morning's lesson (Could be one of the questions above or something else!). Use an example in your explanation.
+1. Share one resource you found useful while researching.
+
+
+## Final Checkpoint

```

Comment:
I think this is a good opportunity for some whole-group discussion.  The difference and use-cases for Inheritance and Interfaces can be very nuanced (at least that's how it seems to me).

Reply:
Good call. I'll make an instructor note to run this think/pair/share style so it's more of a discussion than just a slack waterfall or something like that.